### PR TITLE
Product supplier_reference is not updated when saving a product in the BO

### LINKED ISSUE #20094: Product default supplier reference is not updated
#### Describe the bug

When updating a product, supplier_reference attribute should be updated with the reference from default supplier.

This was fixed by https://github.com/PrestaShop/PrestaShop/pull/11175 last year, but broken by https://github.com/PrestaShop/PrestaShop/commit/29bda43a552316fb4a29c60f44ff287230bce62f#diff-7d7493bae4eba5d548b7f7f88d16d7ea

This field is used by some ERP connectors.

@PierreRambaud what do you think ?

#### Expected behavior

Product supplier reference should be updated

#### Steps to Reproduce

Create a product, add a product supplier with reference, edit supplier reference.

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Mark `wholesale_price` and `supplier_reference` as deprecated. These values are already register with `ProductSupplier` ObjectModel, it's useless to duplicate values. I also restore the registration of these fields when there is no combination even if values can be fetched by a request into `ProductSupplier` table.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | yes
| Fixed ticket? | Fixes #20094
| How to test?  | Follow ticket instruction. Make sure the issue #16353 is still resolved.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20198)

### CODE REVIEW

**Review by atomiix — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]
You did a little cleaning with your bugfix 😅  but it seems OK 👍  approving

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

Following these steps, the supplier_reference is not updated

1. Go to the BO > Catalog > Brands & Suppliers > Supplier > Create a new supplier `supplier1` & enable it
2. Go to Products > Add a product, enter the product name
3. Click on Options tab, select the supplier created before
4. Enter the supplier `supplier1` reference & save
5. Go to the BO > Catalog > Brands & Suppliers > Supplier > delete the last supplier `supplier1` created
6. Edit the product > save
7. Go to database, table ps_product in `supplier_reference` is not updated, the `id_supplier` is well updated **0**

I attached a screen record
https://drive.google.com/file/d/17vKlRh8LSWOd0KiQvWFu53mdLD-TIpaN/view?usp=sharing

Thanks!

**Comment by PierreRambaud** [comment]
> Hi @PierreRambaud,
> 
> Following these steps, the supplier_reference is not updated
> 
> 1. Go to the BO > Catalog > Brands & Suppliers > Supplier > Create a new supplier `supplier1` & enable it
> 2. Go to Products > Add a product, enter the product name
> 3. Click on Options tab, select the supplier created before
> 4. Enter the supplier `supplier1` reference & save
> 5. Go to the BO > Catalog > Brands & Suppliers > Supplier > delete the last supplier `supplier1` created
> 6. Edit the product > save
> 7. Go to database, table ps_product in `supplier_reference` is not updated, the `id_supplier` is well updated **0**
> 
> I attached a screen record
> https://drive.google.com/file/d/17vKlRh8LSWOd0KiQvWFu53mdLD-TIpaN/view?usp=sharing
> 
> Thanks!

Not sure it's related to this pull request since I didn't touche the supplier deletion :sweat_smile: 
Can you try with the 1.7.6 please?

**Comment by khouloudbelguith** [comment]
@PierreRambaud, with PS1.7.6.7, the supplier_reference is always empty, it is not updated in the Database.
image
image

Thanks!

**Comment by PierreRambaud** [comment]
> @PierreRambaud, with PS1.7.6.7, the supplier_reference is always empty, it is not updated in the Database.
> image
> image
> 
> Thanks!

Could you have a check on the 1.7.6.0? :sweat_smile: (sorry)

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

Yes, I have the same behavior as PS1.7.6.0.
image



> @PierreRambaud, with PS1.7.6.7, the supplier_reference is always empty, it is not updated in the Database.
> image
> image
> 
> Thanks!

With 1767, it is a regression, but I think it will be fixed with this PR, isn't it?

Thanks!

**Comment by PierreRambaud** [comment]
> Hi @PierreRambaud,
> 
> Yes, I have the same behavior as PS1.7.6.0.
> image
> 
> > @PierreRambaud, with PS1.7.6.7, the supplier_reference is always empty, it is not updated in the Database.
> > image
> > image
> > Thanks!
> 
> With 1767, it is a regression, but I think it will be fixed with this PR, isn't it?
> 
> Thanks!

No, this pull request is only here to restore the old behavior which means adding the value in supplier_reference field. Everything will be removed in the next major to use only the ProductSupplier Object Model.

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

I found another issue,
In fact, it is related to this issue: https://github.com/PrestaShop/PrestaShop/issues/16353
1. Create 3 suppliers
2. Create a product with
2.1 Price = 12 tax included / =10 tax excluded
2.2 Tax =20%
2.3 Price per unit= 10 /ml
2.4 Cost price = 10
2.5 In the Options Tab, select one supplier as a default.
2.6 Price (tax excl.) for the supplier = 10 
3. Enable product & Save
4. Go to the Catalog page
5. Edit the product
6. change the Price (tax excl.) for the supplier = 8
6. Click on the Go to catalog page
7. Check this product again & check the cost price became 8.
8. if you try to change the cost price & save => green alert is displayed => but it is always saved as 8

https://drive.google.com/file/d/19DH4fZ69tZqbAabTTX8m3e4d-tSWDgYt/view?usp=sharing

Thanks!

**Comment by matks** [comment]
Hi @khouloudbelguith,

I think things are bit mixed up in this thread, because there are 3 code references:
- code (and behavior) on 1.7.6.7
- code on 1.7.7.x
- code on @PierreRambaud PR

## About this scenario

> 1. Go to the BO > Catalog > Brands & Suppliers > Supplier > Create a new supplier `supplier1` & enable it
> 2. Go to Products > Add a product, enter the product name
> 3. Click on Options tab, select the supplier created before
> 4. Enter the supplier `supplier1` reference & save
> 5. Go to the BO > Catalog > Brands & Suppliers > Supplier > delete the last supplier `supplier1` created
> 6. Edit the product > save
> 7. Go to database, table ps_product in `supplier_reference` is not updated, the `id_supplier` is well updated **0**

I analyzed the database. I think field `id_supplier` is wrong on 1.7.7.x and @PierreRambaud PR does not modify that.

The field `id_supplier` is filled when a supplier is linked to a product **and set as the "default supplier"** for this product.
- If you create a product and link it with a supplier, then save ; id_supplier is zero
- If you create a product and link it with a supplier, then mark it as default supplier and save ; id_supplier is OK

When I create a product and link it with a supplier, then mark it as default supplier and save, my id_supplier is OK.
**But then I delete this supplier** it is not updated (on 1.7.7.x). Which is not what you did.

Can you check again your scenario ? Maybe you mixed the 2 scenarios "with default supplier" or "without default supplier". I think in your case id_supplier was always zero.

**Comment by khouloudbelguith** [comment]
Hi @matks,

Thanks for your feedback.
>If you create a product and link it with a supplier, then save ; id_supplier is zero

In fact, this is an old old issue reported here: https://github.com/PrestaShop/PrestaShop/issues/9580
I didn't check this case because w need always to select a supplier by default.

>When I create a product and link it with a supplier, then mark it as default supplier and save, my id_supplier is OK.
But then I delete this supplier it is not updated (on 1.7.7.x). Which is not what you did.

>Can you check again your scenario ? Maybe you mixed the 2 scenarios "with default supplier" or "without default supplier". I think in your case id_supplier was always zero.


There is a small difference between the steps that I mentioned.
In fact, after deleting the supplier, I check the product and I save it again.
As mentioned in the steps 6.

We have two possibilities:
1. If we save a product after deleting a supplier, the id_supplier is updated but the reference_supplier not updated
2. If we don't save a product after deleting a supplier, the id_supplier is not updated and the reference_supplier not updated.

I attached a screen record
https://drive.google.com/file/d/1BPtOq01xPeWta59uSsoUoWRPx_OYgiv1/view?usp=sharing
Thanks!

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]
@khouloudbelguith Aaaaah now I see 😄  thanks ! Indeed it was a very accurate scenario.

@PierreRambaud added some bugfixes, it should be OK now

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

Sorry, but I found two issues.
I just re-install the shop with this **PR**

First issue: `supplier_reference` is not saved at the **first** save

1. Create 3 suppliers
2. Create a simple product without selecting a supplier
3. Edit this product
4. Click on Options tab, select the supplier and mark it as a default
5. Enter the supplier supplier1 reference & save
6. Check the Database 
7. See error: id_supplier is well updated but the `supplier_reference` is empty
8. If you save the product again => it will be ok
https://drive.google.com/file/d/1NFZcn2peCTGuHWmEguT-8hpzL6Dr1lQO/view?usp=sharing


Second issue:
it is related to this issue: https://github.com/PrestaShop/PrestaShop/issues/16353
1. Create 3 suppliers
2. Create a product with
2.1 Price = 12 tax included / =10 tax excluded
2.2 Tax =20%
2.3 Price per unit= 10 /ml
2.4 Cost price = 10
2.5 In the Options Tab, select one supplier as a default.
2.6 Price (tax excl.) for the supplier = 10 
3. Enable product & Save
4. Go to the Catalog page
5. Edit the product
6. change the Price (tax excl.) for the supplier = 8
6. Click on the Go to catalog page
7. Check this product again & check the cost price became 8.
8. if you try to change the cost price & save => green alert is displayed => but it is always saved as 8
https://drive.google.com/file/d/1v3Eufda2RPUdi2NTGDM-64vIAvvAuOOM/view?usp=sharing

Thanks!

**Review by matks — DISMISSED** [review]

File: classes/Product.php [hunk 1]
```diff
@@ -7159,4 +7163,33 @@ public function softDeleteCustomizationFields($customizationIds)
 
         return $return;
     }
+
+    /**
+     * Update default supplier data
+     *
+     * @param int $idProduct
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
     * @param int $idProduct
     * @param int $idSupplier
```

**Review by Progi1984 — DISMISSED** [review]

File: classes/Product.php [hunk 2]
```diff
@@ -7159,4 +7163,34 @@ public function softDeleteCustomizationFields($customizationIds)
...
+     * @param int $idSupplier
+     * @param float $wholesalePrice
+     * @param string $supplierReference
+     *
+     * @return bool
+     */
+    public function updateDefaultSupplierData(int $idProduct, int $idSupplier, string $supplierReference, float $wholesalePrice): bool
```
Comment (jolelievre) [inline classes/Product.php]: Since `$this` is not used maybe it should be a static method
Or else you should use `$this->id` instead of the `$idProduct` parameter

Comment (PierreRambaud) [reply classes/Product.php]: Thank you're right, I added the $idProduct after for a test and keep it :sweat_smile:

**Comment by matks** [comment]
I checked the latest changes on this PR and, for my test, it solves both issues reported here https://github.com/PrestaShop/PrestaShop/pull/20198#issuecomment-661721862

**Comment by khouloudbelguith** [comment]
Hi @matks,

In fact, I found this issue after last changes:

1. create a product
2. set Price (tax incl.) =12  // Price per unit= 10/ml   //  cost price =10
3. enable product & save
4. generate two combinations
5. edit combinations with bulk action  //   Cost Price =9  // impact on price =2  //  **Availability date => 2020-07-24**
6. save
7. select a supplier as a default
8. save
9. Check the product again, the default option is not updated

**Without PR**: https://drive.google.com/file/d/1bTGdEfB__Ys_Bk3IQRZKUeLRMoPAmnCW/view?usp=sharing
**With PR**: https://drive.google.com/file/d/1Hhv6SivAeAGoYt2-Hy0nKrps2ChqZeJV/view?usp=sharing

Thanks!

**Comment by matks** [comment]
> Hi @matks,
> 
> In fact, I found this issue after last changes:
> 
> 1. create a product
> 2. set Price (tax incl.) =12  // Price per unit= 10/ml   //  cost price =10
> 3. enable product & save
> 4. generate two combinations
> 5. edit combinations with bulk action  //   Cost Price =9  // impact on price =2  //  **Availability date => 2020-07-24**
> 6. save
> 7. select a supplier as a default
> 8. save
> 9. Check the product again, the default option is not updated
> 
> **Without PR**: https://drive.google.com/file/d/1bTGdEfB__Ys_Bk3IQRZKUeLRMoPAmnCW/view?usp=sharing
> **With PR**: https://drive.google.com/file/d/1Hhv6SivAeAGoYt2-Hy0nKrps2ChqZeJV/view?usp=sharing
> 
> Thanks!

Thanks ! I am currently exploring this bug

I have reproduced the bug with a shorter scenario:
1. create a product
2. generate two combinations
3. edit "all combinations" with bulk action : set Cost Price to 9  and Availability date to 2020-07-24
4. link product to a supplier and mark it as a default
5. save, refresh page
6. see the "default" setting is gone

**Comment by PierreRambaud** [comment]
I found why the problem occurred.
This was because of the `new Product` when setting the `available date`. This one was using cache and retrieved the old data. The same problem spotted when saving the supplier_reference (which was erased the Product Cost). The only "proper" way I found without having to refactor the whole AdminProductWrapper was to disable cache before loading the Product and reenable it just after.

**Review by Progi1984 — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

I found this issue: In fact, it is related to this issue: #16353

1. Create 3 suppliers
2. Create a product with
2.1 Price = 12 tax included / =10 tax excluded
2.2 Tax =20%
2.3 Price per unit= 10 /ml
2.4 Cost price = 10
=> check the table `ps_product,` you will find that **wholesale_price** = 0 => NOK
2.5 In the Options Tab, select one supplier as a default.
2.6 Price (tax excl.) for the supplier = 5
3. Enable product & Save
4. Go to the Catalog page

=> check the table `ps_product,` you will find that **wholesale_price** = 5 => NOK
[Product • pr20198.webm.zip](https://github.com/PrestaShop/PrestaShop/files/4989486/Product.pr20198.webm.zip)


If we have a product with combination => issue not reproduced => OK.

I checked the Branch 1.7.7.x => ok
`wholesale_price = cost price`

Thanks!

**Comment by PierreRambaud** [comment]
> * Price (tax excl.) for the supplier = 5

The `wholesale_price` is the  `Price (tax excl.) for the supplier = 5`, it's the copy from the ProductSupplier object model, it's defined when you define a supplier without combination like in my screenshot :thinking:
image

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

There is a difference between **with/without** PR 
**Without PR**
You have already suppliers
1. Create a product
2. Set `cost price = 5`
3. Save
4. Check the database => ps_product => `wholesale_price= cost price = 5`
5. Go back to the product => Options Tab
6. Select a supplier an,d set a `price` for this supplier = `10`
7. Check the database => ps_product => `wholesale_price= cost price = 5`

[Without PR.webm.zip](https://github.com/PrestaShop/PrestaShop/files/4993520/Without.PR.webm.zip)

**With PR**
You have already suppliers
1. Create a product
2. Set `cost price = 5`
3. Save
4. Check the database => ps_product => `wholesale_price= cost price = 0`
5. Go back to the product => Options Tab
6. Select a supplier an,d set a `price` for this supplier = `10`
7. Check the database => ps_product => `wholesale_price= cost price = 10`

[With PR.webm.zip](https://github.com/PrestaShop/PrestaShop/files/4993527/With.PR.webm.zip)


Thanks!

**Comment by jolelievre** [comment]
@khouloudbelguith I think it makes sense to update this field and keep it in sync with the ProductSupplier data

In the product we store the default supplier data, it's a kind of shortcut, or some legacy fields that were kept for backward compatibility It was maybe not updated before but it was an error I think

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

Thanks for your feedback.
I think we should ping the Product team to get a specifications about this issue.
Ping @PrestaShop/prestashop-product-team what do you think about this [behavior](https://github.com/PrestaShop/PrestaShop/pull/20198#issuecomment-665509514) ?

Thanks!

**Comment by marionf** [comment]
I think cost price is used when there is no suppliers and it make sense if the cost price takes the same value as the default supplier price (otherwise, I don't see what the default supplier will be for :thinking:)
@simondaigre, as you seems to use suppliers feature, does this behavior make sense for you ?

**Comment by simondaigre** [comment]
@marionf IMHO (I manage around 100 PS instances), few of them use PS suppliers because they manage this with their ERP, and managing suppliers with PS **and** their ERP would be duplicated work for merchants, and many of them doesn't want to.

However, these two fields are used by every all of them. `supplier_reference` is mainly used by ERP connectors for syncing, `wholesale_price` is used for PS stats, marketing/retargeting modules and modules who re-position products according to "commercial margin" (I'm not sure about this term).

**TLDR**: `supplier_reference` and `wholesale_price` is really useful for merchants who doesn't manage their suppliers with PS.

**Comment by marionf** [comment]
Specs updated with the correct behavior: https://github.com/PrestaShop/prestashop-specs/commit/e51eb3217f6ddc8f5e3a5c52eeab0cf1a36cdc62

**Comment by khouloudbelguith** [comment]
Hi @marionf,

Thanks!
@PierreRambaud, so as detailed here: https://github.com/PrestaShop/prestashop-specs/commit/e51eb3217f6ddc8f5e3a5c52eeab0cf1a36cdc62
Following these steps:

1. Create a product
2. Set cost price = 5
3. Save
4. Check the database => ps_product => wholesale_price= cost price = 0

There is no supplier selected and the cost price is always **0** => this is incorrect.

Thanks!

**Comment by PierreRambaud** [comment]
> Hi @marionf,
> 
> Thanks!
> @PierreRambaud, so as detailed here: [PrestaShop/prestashop-specs@e51eb32](https://github.com/PrestaShop/prestashop-specs/commit/e51eb3217f6ddc8f5e3a5c52eeab0cf1a36cdc62)
> Following these steps:
> 
> 1. Create a product
> 2. Set cost price = 5
> 3. Save
> 4. Check the database => ps_product => wholesale_price= cost price = 0
> 
> There is no supplier selected and the cost price is always **0** => this is incorrect.
> 
> Thanks!

Yes, this is because the Cost Price is stored in `ps_product_shop`, the `wholesale_price` is only for the Default Supplier shortcut
image

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

Thanks for these clarifications.
It is ok :heavy_check_mark: 
https://drive.google.com/file/d/1QxtdbypPW7zM_-jG5CVKPDwJMEzS5LVw/view?usp=sharing

Thanks!

**Comment by matks** [comment]
Thank you @PierreRambaud and @khouloudbelguith
